src/Mod/Part/Gui/SectionCutting.cpp

```diff
@@ -0,0 +1,1475 @@
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+# include <Inventor/actions/SoGetBoundingBoxAction.h>
+# include <Inventor/nodes/SoCamera.h>
+# include <Inventor/nodes/SoClipPlane.h>
+# include <Inventor/nodes/SoGroup.h>
+# include <Inventor/nodes/SoOrthographicCamera.h>
+# include <Inventor/sensors/SoTimerSensor.h>
+# include <QDialog>
+# include <QDockWidget>
+# include <QPointer>
+//# include <cmath>
+#endif
+
+#include "SectionCutting.h"
+#include "ui_SectionCutting.h"
+#include <App/Link.h>
+#include <App/Part.h>
+#include <Base/UnitsApi.h>
+#include <Gui/Application.h>
+#include <Gui/Command.h>
+#include <Gui/DockWindowManager.h>
+#include <Gui/Document.h>
+#include <Gui/MainWindow.h>
+#include <Gui/View3DInventor.h>
+#include <Gui/View3DInventorViewer.h>
+#include <Mod/Part/App/FeatureCompound.h>
+#include <Mod/Part/App/FeaturePartBox.h>
+#include <Mod/Part/App/FeaturePartCommon.h>
+#include <Mod/Part/App/FeaturePartCut.h>
+#include <Mod/Part/App/FeaturePartFuse.h>
+#include <Mod/Part/App/PartFeatures.h>
+
+using namespace PartGui;
+
+enum Refresh : bool
+{
+    notXValue = false,
+    notYValue = false,
+    notZValue = false,
+    notXRange = false,
+    notYRange = false,
+    notZRange = false,
+    XValue = true,
+    YValue = true,
+    ZValue = true,
+    XRange = true,
+    YRange = true,
+    ZRange = true
+};
+
+SectionCut::SectionCut(Gui::View3DInventor* view, QWidget* parent)
+    : QDialog(parent)
+    , ui(new Ui_SectionCut)
+{
+    // create widgets
+    ui->setupUi(this);
+    ui->cutX->setRange(-INT_MAX, INT_MAX);
+    ui->cutY->setRange(-INT_MAX, INT_MAX);
+    ui->cutZ->setRange(-INT_MAX, INT_MAX);
+
+    view = view;
+    Gui::View3DInventorViewer* viewer = view->getViewer();
+
+    // get all objects in the document
+    doc = Gui::Application::Instance->activeDocument()->getDocument();
+    if (!doc) {
+        Base::Console().Error("SectionCut error: there is no document\n");
+        return;
+    }
+
+    std::vector<App::DocumentObject*> ObjectsList = doc->getObjects();
+    if (ObjectsList.empty()) {
+        Base::Console().Error("SectionCut error: there are no objects in the document\n");
+        return;
+    }
+    // empty the ObjectsListVisible
+    ObjectsListVisible.clear();
+    // now store those that are currently visible
+    for (auto it = ObjectsList.begin(); it != ObjectsList.end(); ++it) {
+        if ((*it)->Visibility.getValue())
+            ObjectsListVisible.push_back((*it));
+    }
+
+    // we can have existing cut boxes take their values
+    // the flip state cannot be readout of the box position, therefore readout the position
+    // is if it was unflipped
+    if (doc->getObject(BoxZName)) {
+        hasBoxZ = true;
+        ui->groupBoxZ->setChecked(true);
+        Part::Box* pcBox = static_cast<Part::Box*>(doc->getObject(BoxZName));

```

Comment:
But in the next line I check if it is a Part::Box. What is my mistake?

Reply:
You assume that an existing object "SectionCutBoxZ" must be a Part::Box because you create an object in the dialog with this (default) name. But you ignore the fact that the user may have created an object with this name that is not a Part::Box but e.g. a Mesh::Cube.

Doing a static_cast on a pointer is _not_ a type check. The static_cast assumes that you know what you are doing and simply returns a pointer to a Part::Box. In case the object is a Part::Box it works but silently fails if it's not. For the latter case you still get a pointer but you cannot trust it because it's already a dangling pointer.

There are two way to solve this:
* replace the static_cast with a dynamic_cast. This cast internally checks the class' typeid and if it doesn't match it will return a nullptr.
* before doing a static_cast you must use the function `isDerivedFrom(Part::Box::getClassTypeId())`. Internally this uses FreeCAD's implementation of type ids.

Reply:
> Doing a static_cast on a pointer is not a type check.

Aha! Lesson learned. I thought this is the case.
@chennes since you also had a look at this code, it might be a valuable info for you too.

Reply:
done in #6360